### DISCUSSION TITLE
Error in Executing the code

### DISCUSSION THREAD

**Discussion by Dabs212** [alan-turing-institute/sktime#3941] [General]
I am trying to predict FF_Rate as a function of other variables in the Excel Sheet.
Here is the code

import warnings
import numpy as np
import pandas as pd
import sktime
import datetime
from datetime import datetime
from openpyxl import Workbook, load_workbook
from sktime.registry import all_estimators
from sktime.datasets import load_airline
from sktime.utils.plotting import plot_series
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import make_reduction
# from sktime.forecasting.compose import ReducedRegressionForecaster
from sktime.utils.plotting import plot_series
#from sktime.utils.plotting.forecasting import plot_series
from sklearn.svm import SVR
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sktime.forecasting.naive import NaiveForecaster
from sklearn.ensemble import GradientBoostingRegressor
from sktime.performance_metrics.forecasting import MeanAbsolutePercentageError 
from sktime.forecasting.var import VAR
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.datasets import load_airline
import matplotlib.pyplot as plt
# hide warnings
warnings.filterwarnings("ignore")
#          
#return a complete list of estimators as pd.Dataframe
#
print("All Estimators: ",all_estimators(as_dataframe=True))
all_estimators(as_dataframe=True)
# return all forecasters by filtering for estimator type
print("All Forecasters: ",all_estimators("forecaster"))
# return all forecasters which handle missing data in the input by tag filtering
print(all_estimators("forecaster", filter_tags={"handles-missing-data": True}))
#
# Step-1 get Data
#
file = r"C:\\Daily_Stock_Report\\FRED\\Fred_Mnthlylin_1980.xlsx"
sname = "Master"
#
dg = pd.read_excel(file,sheet_name=sname)
dg = dg.iloc[:,1:] #Ignore 1st column containing index
#
#
dlist = ["EFF_Rate","T3MO",'10YM1Y', '1YrM3Mo', '30YM10Y', '30YM15Y', 'AAA', 'BBB','CrckSprd11', 'CrckSprd11CpG', 
         'CrckSprd11DpB','CrckSprd321', 'CrckSprd321CpG', 'CrckSprd321DpB','NGLMHH', 'CrckSprd532CpG',
    'CrckSprd532CpG','CrckSprd6321', 'CrckSprd6321CpG', 'Brent', 
    'CrckSprd6321DpB','BrentMWTI','NGLMHH','PPI_IG','PPI_FG','Fertilizers','Bus_Inventory',
     'R_PCEDG', 'R_PCENDG', 'R_PCESC', 'RbrPlas', 'T10Y2Y','Propane', 'Ethanol','GsolnMDiesel',
        'JetMGasoln', 'JetMDiesel', 'G_NatGasMHH', 'DieselMFO','OneYr', 'WGS10YR', 'MORT15US','GNATGAS',
         'STF_Stress', 'Job_Opnings', 'Labor_Particip', 'Veh._Purchs', 'CrckSprd532DpB',
         'RES_Balnc', 'CBLiqSwaps', 'OtherLiqDepos', 'LiqSwapsM2wks', 'FuelUtil', 'Copper',
        'Gasoline', 'Diesel', 'JetFuel', 'FuelOil','SahmRule', 'Prob_Recesn','EURO2DLR', 'PND2DLR', 
         'Food', 'PPI_CG', 'Indus_Prodn', 'CapacityUtil', 'Bus_Loans', 'DurG_Orders', 'P_Savings',
         'REM2DLR','YEN2DLR','NASDAQ100', 'DJIA','SP500','BldgPermits']
df = pd.DataFrame(dg.drop(dlist,inplace=False, axis=1))
df.bfill(axis='rows')
df.ffill(axis='rows')
df
list_df_col = df.columns.values.tolist()
print(list_df_col)
#
fh = np.arange(1, 10)
forecaster = NaiveForecaster(strategy="last", sp=12)
#
FF_Rate = df["FF_Rate"]
y = FF_Rate
#
X = df["Inflation"]
X = X.append(df["Unemployment"])
print(X)
y = FF_Rate
#
#step 4: fitting the forecaster
#
forecaster.fit(y, X=X)
#
# step 5: querying predictions
y_pred = forecaster.predict(fh=fh, X=X)
#
plot_series(y_pred, y, labels = ["Prediction", "Actual"])

It gives follow error:
TypeError                                 Traceback (most recent call last)
Cell In[1], line 81
     77 y = FF_Rate
     78 #
     79 #step 4: fitting the forecaster
     80 #
---> 81 forecaster.fit(y, X=X)
     82 #
     83 # step 5: querying predictions
     84 y_pred = forecaster.predict(fh=fh, X=X)

File ~\AppData\Roaming\Python\Python38\site-packages\sktime\forecasting\base\_base.py:334, in BaseForecaster.fit(self, y, X, fh)
    331 self.reset()
    333 # check and convert X/y
--> 334 X_inner, y_inner = self._check_X_y(X=X, y=y)
    336 # set internal X/y to the new X/y
    337 # this also updates cutoff from y
    338 self._update_y_X(y_inner, X_inner)

File ~\AppData\Roaming\Python\Python38\site-packages\sktime\forecasting\base\_base.py:1385, in BaseForecaster._check_X_y(self, X, y)
   1372 msg = (
   1373     "X must be either None, or in an sktime compatible format, "
   1374     "of scitype Series, Panel or Hierarchical, "
   (...)
   1382     "Possible mtype specification strings are as follows. "
   1383 )
   1384 if not X_valid:
-> 1385     raise TypeError(msg + mtypes_msg)
   1387 X_scitype = X_metadata["scitype"]
   1388 X_requires_vectorization = X_scitype not in X_inner_scitype

TypeError: X must be either None, or in an sktime compatible format, of scitype Series, Panel or Hierarchical, for instance a pandas.DataFrame with sktime compatible time indices, or with MultiIndex and last(-1) level an sktime compatible time index. See the forecasting tutorial examples/01_forecasting.ipynb, or the data format tutorial examples/AA_datatypes_and_datasets.ipynbIf you think X is already in an sktime supported input format, run sktime.datatypes.check_raise(X, mtype) to diagnose the error, where mtype is the string of the type specification you want for X. Possible mtype specification strings are as follows. "For Hierarchical scitype: ['pd_multiindex_hier']. 
[Fred_Mnthlylin_1980.xlsx](https://github.com/sktime/sktime/files/10229902/Fred_Mnthlylin_1980.xlsx)

**Comment by Dabs212** [comment]
I still have not received any suggestion on formatting X, it is just like column "FF_Rate" for y. Can anyone help please?

**Comment by miraep8** [comment]
Hello!

The code as you have pasted it honestly makes it a bit hard to parse.  In the future, posting minimal code in code blocks with 3 ` symbols, ie like this:

```
import example_package

X = module(params)....
```
Makes it a bit easier.  Also focusing on a minimally reproducible error (ie what is the smallest amount of code I can use which raises the same error/issue, with a data set we have access to (ie - we do not have access to the private file on your computer)) is also extremely helpful for us to find what might be causing you issues.   I just make these suggestions so that in the future it is easier for you to get the help you would like. 

As to the error you are receiving, certain estimators (our name for the different models you might want to use require different inputs.  My suggestion for how to approach this problem would be to first a) - verify on your end what is the type of `X`?  then check out the documentation on the particular forecaster which is causing you this error and you can see what it expects `X` to look like.   The error message is a good place to start - though perhaps this error message was for a different forecaster?  Just a bit surprised personally to see an error about hierarchical data being raised for `NaiveForecaster`? (can you confirm that this was where you saw the issue?).  Its also worth exploring more on your end - what is the exact problem you are trying to solve? Sometimes these type errors can arise in situations where you may be trying to use an estimator that does not make sense for the problem at hand.

**Reply by lmmentel** [reply]
A good reference for how to share code examples
 is on stack overflow https://stackoverflow.com/help/minimal-reproducible-example

**Comment by Dabs212** [comment]
import warnings
import sys
import numpy as np
import pandas as pd
import sktime
import datetime
import matplotlib.pyplot as plt
import seaborn as sns
import xgboost as xgb       # Had to install it separately
from sklearn.metrics import mean_squared_error
color_pal = sns.color_palette()
plt.style.use('fivethirtyeight')
from datetime import datetime
from openpyxl import Workbook, load_workbook
from sktime.registry import all_estimators
from sktime.datasets import load_airline
from sktime.utils.plotting import plot_series
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import make_reduction
from sklearn.svm import SVR
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sktime.forecasting.naive import NaiveForecaster
from sklearn.ensemble import GradientBoostingRegressor
from sktime.performance_metrics.forecasting import MeanAbsolutePercentageError 
from sktime.forecasting.var import VAR
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.datasets import load_airline
warnings.filterwarnings("ignore")
#          
#return a complete list of estimators as pd.Dataframe
#
print("All Estimators: ",all_estimators(as_dataframe=True))
all_estimators(as_dataframe=True)
# return all forecasters by filtering for estimator type
print("All Forecasters: ",all_estimators("forecaster"))
# return all forecasters which handle missing data in the input by tag filtering
print(all_estimators("forecaster", filter_tags={"handles-missing-data": True}))
#
# Step-1 get Data
#
file = r"C:\\Daily_Stock_Report\\FRED\\Fred_Mnthlylin_1980.xlsx"
sname = "Master"
#
dg = pd.read_excel(file,sheet_name=sname)
#dg = dg.iloc[:,1:] #Ignore 1st column containing index
dg = dg.set_index('Date')
dg.index = pd.to_datetime(dg.index)
print(dg.head())
#
#
dlist = ["EFF_Rate","T3MO",'10YM1Y', '1YrM3Mo', '30YM10Y', '30YM15Y', 'AAA', 'BBB','CrckSprd11', 'CrckSprd11CpG', 
         'CrckSprd11DpB','CrckSprd321', 'CrckSprd321CpG', 'CrckSprd321DpB','NGLMHH', 'CrckSprd532CpG',
    'CrckSprd532CpG','CrckSprd6321', 'CrckSprd6321CpG', 'Brent', 
    'CrckSprd6321DpB','BrentMWTI','NGLMHH','PPI_IG','PPI_FG','Fertilizers','Bus_Inventory',
     'R_PCEDG', 'R_PCENDG', 'R_PCESC', 'RbrPlas', 'T10Y2Y','Propane', 'Ethanol','GsolnMDiesel',
        'JetMGasoln', 'JetMDiesel', 'G_NatGasMHH', 'DieselMFO','OneYr', 'WGS10YR', 'MORT15US','GNATGAS',
         'STF_Stress', 'Job_Opnings', 'Labor_Particip', 'Veh._Purchs', 'CrckSprd532DpB',
         'RES_Balnc', 'CBLiqSwaps', 'OtherLiqDepos', 'LiqSwapsM2wks', 'FuelUtil', 'Copper',
        'Gasoline', 'Diesel', 'JetFuel', 'FuelOil','SahmRule', 'Prob_Recesn','EURO2DLR', 'PND2DLR', 
         'Food', 'PPI_CG', 'Indus_Prodn', 'CapacityUtil', 'Bus_Loans', 'DurG_Orders', 'P_Savings',
         'REM2DLR','YEN2DLR','NASDAQ100', 'DJIA','SP500','BldgPermits']
print(dlist.sort())
df = pd.DataFrame(dg.drop(dlist,inplace=False, axis=1))
df.bfill(axis='rows')
df.ffill(axis='rows')
df
list_df_col = df.columns.values.tolist()
print(list_df_col)
#
# fh = np.arange(1, 10)
# forecaster = NaiveForecaster(strategy="last", sp=12)
#
FF_Rate = df["FF_Rate"]
y = FF_Rate

Sorry, I had line numbers in previous code
Step-1: I am unable to see results of Print Statements (why?)
Step-2 I am trying to predict Feed Funds Rate as a function of Inflation and Unemployment. I submitted my dataset Fred_Mnthlylin_1980.xlsx but does not show up as repository. I am attaching spreadsheet here
[Fred_Mnthlylin_1980_VRD.xlsx](https://github.com/sktime/sktime/files/10255598/Fred_Mnthlylin_1980_VRD.xlsx)

Next item is Forecaster

#
#X = df["Inflation"]
X = pd.concat([df["Inflation"],df["Unemployment"]],axis=1)
print(y,X)
#
#step 4: fitting the forecaster
#
forecaster.fit(y, X=X, fh=fh)
#
# step 5: querying predictions
#
y_pred = forecaster.predict(fh=fh, X=X)
# #
plot_series(y_pred, y, labels = ["Prediction", "Actual"])
#
It throws Error rather long
TypeError                                 Traceback (most recent call last)
Cell In[1], line 93
     89 forecaster.fit(y, X=X, fh=fh)
     90 #
     91 # step 5: querying predictions
     92 #
---> 93 y_pred = forecaster.predict(fh=fh, X=X)
     94 # #
     95 plot_series(y_pred, y, labels = ["Prediction", "Actual"])

File ~\AppData\Roaming\Python\Python38\site-packages\sktime\forecasting\base\_base.py:404, in BaseForecaster.predict(self, fh, X)
    402 # we call the ordinary _predict if no looping/vectorization needed
    403 if not self._is_vectorized:
--> 404     y_pred = self._predict(fh=fh, X=X_inner)
    405 else:
    406     # otherwise we call the vectorized version of predict
    407     y_pred = self._vectorize("predict", X=X_inner, fh=fh)

File ~\AppData\Roaming\Python\Python38\site-packages\sktime\forecasting\naive.py:336, in NaiveForecaster._predict(self, fh, X)
    326 def _predict(self, fh=None, X=None):
    327     """Forecast time series at future horizon.
    328 
    329     Parameters
   (...)
    334         Exogenous time series
    335     """
--> 336     y_pred = super(NaiveForecaster, self)._predict(fh=fh, X=X)
    338     # test_predict_time_index_in_sample_full[ForecastingPipeline-0-int-int-True]
    339     #   causes a pd.DataFrame to appear as y_pred, which upsets the next lines
    340     #   reasons are unclear, this is coming from the _BaseWindowForecaster
    341     # todo: investigate this
    342     if isinstance(y_pred, pd.DataFrame):

File ~\AppData\Roaming\Python\Python38\site-packages\sktime\forecasting\base\_sktime.py:31, in _BaseWindowForecaster._predict(self, fh, X)
     29 # all values are out-of-sample
     30 if fh.is_all_out_of_sample(self.cutoff):
---> 31     return self._predict_fixed_cutoff(
     32         fh.to_out_of_sample(self.cutoff), **kwargs
     33     )
     35 # all values are in-sample
     36 elif fh.is_all_in_sample(self.cutoff):

File ~\AppData\Roaming\Python\Python38\site-packages\sktime\forecasting\base\_sktime.py:65, in _BaseWindowForecaster._predict_fixed_cutoff(self, fh, X, return_pred_int, alpha)
     50 """Make single-step or multi-step fixed cutoff predictions.
     51 
     52 Parameters
   (...)
     62 y_pred = pd.Series or pd.DataFrame
     63 """
     64 # assert all(fh > 0)
---> 65 y_pred = self._predict_last_window(
     66     fh, X, return_pred_int=return_pred_int, alpha=alpha
     67 )
     68 if isinstance(y_pred, pd.Series) or isinstance(y_pred, pd.DataFrame):
     69     return y_pred

File ~\AppData\Roaming\Python\Python38\site-packages\sktime\forecasting\naive.py:203, in NaiveForecaster._predict_last_window(self, fh, X, return_pred_int, alpha)
    199 def _predict_last_window(
    200     self, fh, X=None, return_pred_int=False, alpha=DEFAULT_ALPHA
    201 ):
    202     """Calculate predictions for use in predict."""
--> 203     last_window, _ = self._get_last_window()
    204     fh = fh.to_relative(self.cutoff)
    206     strategy = self.strategy

File ~\AppData\Roaming\Python\Python38\site-packages\sktime\forecasting\base\_sktime.py:124, in _BaseWindowForecaster._get_last_window(self)
    122 # Get the start and end points of the last window.
    123 cutoff = self._cutoff
--> 124 start = _shift(cutoff, by=-self.window_length_ + 1)
    125 cutoff = cutoff[0]
    127 # Get the last window of the endogenous variable.

File ~\AppData\Roaming\Python\Python38\site-packages\sktime\utils\datetime.py:186, in _shift(x, by, return_index)
    184 # we need to get freq from idx, since pd.Timestamp freq is deprecated
    185 if isinstance(x, pd.Timestamp):
--> 186     by *= idx.freq
    187 return x + by

TypeError: unsupported operand type(s) for *=: 'int' and 'NoneType'

**Comment by fkiraly** [comment]
@Dabs212, for us to help and/or debug effectively, we need a short and self-contained piece of code.

Can you please reduce your code to the minimum so it still produces the issue?
If you can reproduce this with dummy data instead of an excel file that we do not have, that would be much appreciated.

Kindly follow the instructions that @miraep8 and @lmmentel posted.

**Reply by Dabs212** [reply]
Thanks Frank. I do not know what to cut out, but one can copy and paste the lines. I see the issue though, it has line numbers. I lost bunch of files while uninstalling miniconda3. Also I attached data-source XLSX file in my posting.
Thanks again.

**Reply by fkiraly** [reply]
it would be great if you could put your code in a fenced code block, see here:
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

ensure all imports are included, and it runs in a clean jupyter notebook until the error pops up

**Comment by XutongGithub** [comment]
邮件已收到。谢谢！

**Comment by Dabs212** [comment]
Thanks Frank, I will try it
Vivek R. Dabholkar 

    On Thursday, December 29, 2022 at 10:08:46 AM CST, Franz Király ***@***.***> wrote:  
 
 


it would be great if you could put your code in a fenced code block, see here:
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

ensure all imports are included, and it runs in a clean jupyter notebook until the error pops up
